### ISSUE TITLE
When to release?

### ISSUE THREAD

**Issue by torvista** [zencart/zencart#2973]
Is there a milestone that will determine the next release?
a) more people on the forum are starting to report notices and warnings that are aleady fixed in 157 (so time/effort wasted there)
b) an idea of the schedule would help to allocate/focus my meagre resources in completing multiple almost-finished PRs for inclusion. I know I should do them all immediately, but it's just not possible in the real world.

**Comment by scottcwilson** [comment]
We have a December release tradition, don't we? :)

**Comment by torvista** [comment]
NOOO!:scream: 

How about when PRs are approaching zero!!

**Comment by drbyte** [comment]
Much as I'd like to put a release out before end of 2019, it's not going to happen this time.

**Comment by lat9** [comment]
> 
> 
> NOOO!😱
> 
> How about when PRs are approaching zero!!
... and the issues' count.

**Comment by scottcwilson** [comment]
> ... and the issues' count.

Only for critical must-fix issues.  Low priority issues and feature requests will always be there.   Right now there is tremendous value in 1.5.7 that is not available to any user.

**Comment by judygunderson** [comment]
> 
> 
> > ... and the issues' count.
> 
> Only for critical must-fix issues. Low priority issues and feature requests will always be there. Right now there is tremendous value in 1.5.7 that is not available to any user.

I agree Scott!

**Comment by torvista** [comment]
Just running through all the pages with "Report All Errors" on still generates errrors, but I believe it is now required, to set a "clean" baseline of no php notices/warnings from core code on php7.3.
I have achieved that with my 156 but it has taken months of my (non) spare time. chipping away...

**Comment by lat9** [comment]
The last thing that we want to do with zc157 is release it too soon.  Other than the squishing of some of the PHP 7.3/7.4 interoperability issues, I'm not seeing any _major_ functional/feature changes that would warrant a zc157 vs. zc156d.

**Comment by judygunderson** [comment]
> 
> 
> The last thing that we want to do with zc157 is release it too soon. Other than the squishing of some of the PHP 7.3/7.4 interoperability issues, I'm not seeing any _major_ functional/feature changes that would warrant a zc157 vs. zc156d.

Definitely agree, but I am like a kid at a candy store and my Mama just saying no no no.  giggle

What I don't want to see is a 1.5.6d come out and then 2 months later 1.5.7 because that ticks my customers off! :(

**Comment by wiztechinc** [comment]
Please just no.

**Comment by drbyte** [comment]
I'd love to see a list of the most popular plugins that people depend on, particularly in order to ascertain the impact of the v157 changes on those plugins ... because there may be an opportunity to tweak v157 slightly further (ie: notifiers, and some code consolidation) so that updating these plugins after v157 may be easier in the future.

(My gut says that the most-impacted plugins between v156 and v157 are admin-related plugins, since catalog-side has a lotta notifier stuff already in place.)

**Comment by wiztechinc** [comment]
my go to is always image handler. COWOA is next but I've pulled it out of my templates now.  Speeding up sites is a real issue and the image handler makes quick work of the pics. I do know there are folks who hate it (from a hosting point of view) but it's a life saver nowadays. Dual pricing is one that when a business needs it, it's necessary but didn't I see earlier discussion about this?

**Comment by torvista** [comment]
I would say the native image handling is an oxymoron. Image handler or such functionality is beyond essential, in my view, I have 18k images on my site, I can't imagine that without image handler to ease the pain.

"impact of the v157 changes"
What changes?
I use a lot of 157 code in my 156 site where appropiate (mainly admin).
You are implying there are some breaking changes, but I have not noticed any.

**Comment by drbyte** [comment]
Aren't a lot of plugins that make updates to Admin files going to experience complex needs when upgrading? ie: all the bootstrap refactoring, etc.

**Comment by drbyte** [comment]
Image Handler is an actively-maintained plugin, by a team of people constantly watching core ZC changes. I'm not worried about its ability to keep up.

I'm also not convinced it's ideal for core code inclusion: it still contains a lot of legacy functionality that feels like bloat IMO.

On the plus side, core code has dozens of hooks in it to facilitate the needs of the IH plugin so that its installation is relatively painless ... even if it's hard to follow when doing an upgrade to a highly customized site.

**Comment by wiztechinc** [comment]
yep, getting better for sure and being maintained. But essential as far as I'm concerned. Therefore, always install it. Always. Sounds like a core feature to me.

**Comment by drbyte** [comment]
Again, this discussion is focused on things that needed addressing before releasing the next version.

**Comment by torvista** [comment]
IH, yes it does almost indeed plugin...but my argument is that if a mod is considered  **essential** it should stop being a plugin and just be "in".

> I'm also not convinced it's ideal for core code inclusion: it still contains a lot of legacy functionality that feels like bloat IMO.
>actively-maintained plugin

Those comments seem to conflict....

But still, understood, maybe it has a lot of one-off/non zc-stylee code, but
- it works/is proven
- users don't care about the under-the-hood elegance
- it can be fixed/improved anytime

And the alternative is....there IS no alternative!

**Comment by drbyte** [comment]
> maybe it has a lot of one-off/non zc-stylee code, but
> 
> * it works/is proven
> * users don't care about the under-the-hood elegance
> * it can be fixed/improved anytime

While it may seem romantic to include in ZC core, as soon as it is, everyone will assume that the ZC dev team is obligated to support and maintain it. The benefit of having it be a plugin is that anyone can install an update to "just the plugin itself" anytime, regardless of there being a new ZC version. ie: maintenance/release cycles can be largely independent of each other.

More specifically: no, I'm not interested in being the maintainer of that feature. Better to have the community do it. It'll definitely be more robust that way.

> And the alternative is....there IS no alternative!

While you may not like it, the alternative is not using the plugin.

**Comment by drbyte** [comment]
I repeat: Again, this discussion is focused on things that needed addressing before releasing the next version.

**Comment by scottcwilson** [comment]
Two questions:
a) How close is @zcwilt to finishing a music plugin?
b) Should we merge in the functions folders from 2.0 prior to release?

**Comment by drbyte** [comment]
> Should we merge in the functions folders from 2.0 prior to release?

How willing is everyone to break all the functions files that plugins regularly customize?

**Comment by scottcwilson** [comment]
My records (>5 years worth) show 95 plugins that customize functions files.   Not a huge number.

We have done *some* function consolidation in 1.5.7.  We could stop here, keep going, or just import the 2.0 changes (and add the bug fixes that have been done in 1.5 but not 2.0). 

Some of these are abandoned, some are duplicates, some are older versions.   

Add_Shipping_Telephone
Admin_Login_As_Customer
CSAR
CompanyNameAfterCustomerName
Customer_Tax_Exempt
Display_Specific_Categories
Dual-Pricing
Dynamic_Attribute
EmailsFromCustomersInAdmin_v1
Events_Calendar
FAQ_Manager_2
GreyClean
MAP
MSRP
Moosend
Multi-Language-Country-Names
Multi-language_SEO
News_Articles_Management
Personalized_My_Account
ProductsUpREST
Russian
StockByAttributes_1
Testimonial-Manager
Transcendence
TwitchAdditionalProductImages
TwitchWholesalePlusAttributes
Twitch_Restrict_Cat
ZCA-Responsive-Classic
ZCA_BOOTSTRAP
ZCA_BOOTSTRAP_TEMPLATE
add_customers_from_admin_2
add_customers_from_admin_v1
admin_new_customer
admin_password_change_display
admin_profiles_v1_1_1
alac
authorize_net_cim_0
cart_total_weight
ceon_back_in_stock_notifications
ceon_uri_mapping
common_email_styling
creinold_SBA5
css3_buttons
css_buttons_admin
dangling_carrot
dbio
dropship_core
dual_pricing
dynamic_filter
echo_image_load
email_archive_manager
fluorspar
gift_certificate_addon
group_pricing_per_item
group_pricing_unlimited
lazy_load
master_password
minimum_customer_account_information
multiship
multisite
news_box_manager_v3-3
newsletter_subscribe
numinix_barcodes
old_customer_tax_exempt
one_page_checkout
osh_updated_by
product_delivery_by_postcode
qbi
selectable_countries
sendinblue
shift4
simple-seo-url_1
snap_affiliates
sortable_orders_status
stock_by_attribute
stock_by_attributes
store_credit
styleable_minmaxunits
tableau
tableau2
taxcloud
taxcloud_1
testimonial_manager
tw_ws_STD_18c_for_155f
twitch_TURN_OFF_sm_specials_dc
twitch_salemaker_attrs
twitch_wholesale
ty_package_tracker
ultimate_seo_urls
ultimate_seo_urls_2
vat4eu
zen_bck
zjewel_responsive_template
zucando_responsive_fashion
zx_optimized_images

**Comment by wiztechinc** [comment]
Seeing that list makes me panic. The ceon mapping mod is still being used and there’s no way to convert it to any other mod. No fixes, future problems for the moment, bless Conor’s heart. I truly do not know what functions you are talking about but moving too fast can literally drive zen cart users away. Please take care.

**Comment by lankeeyankee** [comment]
edit orders
dbio
IH5 (there are other "image handler" modules that do similar things)
keep cart
structured data
google enhanced ecommerce analytics
info at a glance
print invoice popup
store credit
xml sitemap
ask a question (modified to stay on product info page using accordion or modal script)
ajax back in stock notifications
backup mysql
ckeditor
dgreviews or some other review module
dynamic price updater
google products search feature (commercial, I know, but a very important module if using google paid services)
ceon advanced shipper(commercial, but could not operate without it since zen cart shipping options are all or nothing product-wise and too basic)
quick updates
rma

This is the main list of what I consider essential to basic operation, and I think these should be in the core because my guess is the majority of physical goods sellers would find them all useful and many/most would be useful for any type of ecommerce. 

There are more, but they are more specific to certain needs such as dual pricing.

**Comment by torvista** [comment]
> I repeat: Again, this discussion is focused on things that needed addressing before releasing the next version.
Things that **need** addressing should have PRs pending or in the offing or be in the issues.

I read this to mean you want to do a release. Looking at my orginal post of this thread, I don't see any answer yet.

A new version has bugfixes, code improvements, admin/management improvements and catalog features. In my view unfortunately only the last one has any "pull" for the majority of users (not developers/tech users) to re-enter the upgrade "Room 101".

Unfortunately I think users only upgrade when they are forced to/their shop is broken, or there is something they really want in the latest version...and if they really wanted it, they probably already have it as an addon. This makes it very hard to encourage an upgrade.
I appreciate the underlying fixes, but I am loathe to upgrade....I, like many, have a business to run and every upgrade procedure stops me doing that. Hence I now try and add the tweaks I use/need to the core to reduce the pain next time.

So, I think adding some new features is necessary if only from a psychological point of view/to put something on the **outside** of the tin, otherwise few will bother...so why bother releasing it if the takeup is low.
"What new catalog features are in 157 for 2020?"
 
But new features are added and checked by few developers and time marches on, so it's Catch 22.
No easy answer with so few contributors.

**Comment by torvista** [comment]
> The ceon mapping mod is still being used and there’s no way to convert it to any other mod. No fixes, future problems for the moment, bless Conor’s heart. 

Anyone who appreciates the awesomely-written free part of this mod (and Conor's memory) should look down the back of the sofa to get the loose change that the Mapping Manager costs and get the support (which is great). Same for Advanced Shipper, they are low-cost, no brainer decisions for any business.

**Comment by lat9** [comment]
> 
> 
> I repeat: Again, this discussion is focused on things that needed addressing before releasing the next version.

1. The updated plugin interface is neither complete nor documented.
2.  How about addressing the various rounding issues that exist with tax- and quantity-discount-calculations?  Noting that this one isn't easy, as those calculations are currently distributed among various class, function, header and (arggh) template files (e.g. `tpl_account_history_info_default.php`).

**Comment by dbltoe** [comment]
We have a lifetime unlimited use license for MagicThumb and MagicScroll.  Not sure if inclusion of IH would interfere with it's operation.  Just mentioning as someone who would want to be able to turn off IH.

**Comment by scottcwilson** [comment]
> ... this discussion is focused on things that needed addressing before releasing the next version.

@drbyte Do you feel like we have arrived at an answer?

**Comment by drbyte** [comment]
@ everyone ...
So ... on a scale of 1 to 10, how painful is the 156-to-157 upgrade going to be (compared to 155-to-156) with all the changes everyone has submitted?

**Comment by torvista** [comment]
What is the killer reason (for users) for this upgrade?

**Comment by scottcwilson** [comment]
Error free operation in current versions of PHP.

**Comment by torvista** [comment]
It was a leading question....so it can be assumed that the automated testing opens every file and presses every button/option in admin and catalog and no debugs are generated with Report All Errors on?
The upgrade process is tedious and painful always, so the end result needs to be seen to deliver.
Early adopters reporting notice/warning debugs soon after release would not be very encouraging for the prudent/late adopters.
Given the rapidly-required updates required last time, I think an ask in the forum for interested parties to try a beta pre-release would be prudent.

**Comment by lat9** [comment]
> 
> 
> Error free operation in current versions of PHP.

I'm not sure that that's a _killer reason_.  Most shop-owners would like some additional/improved **_functionality_** rather than a _promise_ of continued operation!

**Comment by scottcwilson** [comment]
We have made a bunch of improvements and fixes.
People who get an upgrade should get these fixes now, rather than having to pay for another upgrade later when PHP 7.2 goes end of life, or to pay to have the issues fixed as they arise one at a time.
The project should aim to release at least once a year.  Annually is a very reasonable cadence.   If you don't release often enough, people start to think your project has been abandoned or has too much internal strife to do updates.   
No one is forcing you to take these releases.  You can wait as long as you want!  But when you're ready to go, it would be nice if the release was there.

**Comment by wiztechinc** [comment]
My 2 best arguments for upgrading are enhanced security and functionality. I mean really now. I'm upgrading 1.3.9 and 1.5.0 carts constantly now that the php levels are changing. So often it's only the threat of the cart ceasing to check out  that spurs many owners. Even when the cart doesn't work optimally on a higher level of php, they'll leave it be unless folks can't buy from them. So I would prefer to see more visible changes in each release. And now folks are having to make decisions such as getting rid of COWOA when upgrading. That's less functionality, not more. And yes I prefer the original COWOA over any thing else I've used.

**Comment by torvista** [comment]
I think this should be touted as a maintenance release.
Although we here know it is much more than than that under the hood, that's irrelevant to shop owners with a happily-working cart who will stick with that until it isn't (working).

**Comment by lankeeyankee** [comment]
I have not tested upgrading a vanilla 1.5.6c to the latest release so I do not know on a scale of 1-10 what the PITA level might be. I recall there being some big changes started, using namespaces and modernizing the code, and possibly the start of a true plugin framework. Will this make it harder to upgrade? I will spend a little time testing over the next few days.

**Comment by scottcwilson** [comment]
Like I was saying, "We made a bunch of improvements and fixes."   If *contributors to the project* can't find them, what hope do ordinary users have?  This is why we should release.
https://github.com/zencart/zencart/issues/3139

**Comment by drbyte** [comment]
I'd like to aim at May 15 as a potential target for v1.5.7 release.
(Worst case might push it as far as May 30 if for some reason there are bugs that need correcting.)

I think it's important to get it out so that all the PHP 7.2 and 7.4 "strict" fixes are packaged so people can confidently use the latest PHP version.

We should be careful to not start adding new things that can't be sufficiently battle-tested before release.

**Comment by scottcwilson** [comment]
Roger that.  Suggest that after 4/30, only CRITICAL fixes be merged in.

**Comment by Zen4All** [comment]
@drbyte 
Thank,for the update. I have one smal, but important, bug fix to do for option value manager

**Comment by scottcwilson** [comment]
@torvista suggest closing this issue, creating a new one for tracking todo's before the release.
